### PR TITLE
feat: wasm & protobuf integration

### PR DESCRIPTION
**Pull request by d-roak**
fixes: #6, #8 and #14

## Features
- [x] protobuf for network messages and object ser/deser
- [x] wasm compilation within the `object` package
- [ ] ~~refactor `crdt` package to support `asc` compilation (#91)~~
- [x] refactor `examples` folder to support the new structure (#91)

### CODE REVIEW

**Comment by d-roak** [comment]
Currently, we have an ugly thing for protobuf where we need to manually append ".js" to the `protobufjs/minimal` import because of our `NodeNext` module resolution.

We should track that this PR will fix that: https://github.com/stephenh/ts-proto/pull/1058
If it doesn't, we need to open an issue/pr to fix it

File: examples/chat/src/objects/chat.ts [hunk 1]
```diff
@@ -1,32 +1,41 @@
...
-    chat: GSet<string>;
-    addMessage(timestamp: string, message: string, node_id: string): void;
-    getMessages(): GSet<string>;
-    merge(other: Chat): void;
-}
+// if it can't compile, append src/index.asc to the import path on runtime
+import { GSet, gset_create, gset_add, gset_merge } from "@topology-foundation/crdt";
+
+export class Chat {
+  // store messages as strings in the format (timestamp, message, peerId)
```
Comment (guiltygyoza) [inline examples/chat/src/objects/chat.ts:5]: `peerID` -> `node_id` for consistency

**Review by guiltygyoza — APPROVED** [review]
big effort

File: packages/crdt/src/crdts/GSet/index.ts [hunk 2]
```diff
@@ -1,28 +1,51 @@
...
+  gset.add(element);
+}
+
+export function gset_lookup<T>(gset: GSet<T>, element: T): boolean {
+  return gset.lookup(element);
+}
+
+export function gset_compare<T>(gset: GSet<T>, peerSet: GSet<T>): boolean {
+  // @ts-ignore
+  return (gset.set.size == peerSet.set.size && gset.set.values().every(value => peerSet.set.has(value)));
```
Comment (h3lio5) [inline packages/crdt/src/crdts/GSet/index.ts:41]: why not do`gset.compare(peerSet)` instead of rewriting the logic?

Comment (d-roak) [reply packages/crdt/src/crdts/GSet/index.ts:41]: asc compiler

File: packages/crdt/src/crdts/GSet/index.ts [hunk 3]
```diff
@@ -1,28 +1,51 @@
...
+export function gset_merge<T>(gset: GSet<T>, peerSet: GSet<T>): void {
+  let set = gset.set.values();
+  // @ts-ignore
+  for (let i = 0, l = peerSet.set.values().length; i < l; ++i) {
+    // @ts-ignore
+    gset.add(set[i]);
   }
-}
\ No newline at end of file
+}
```
[the remark this answers is not in the document]
Comment (d-roak) [reply packages/crdt/src/crdts/GSet/index.ts:51]: asc compiler

**Pull request merged**